libsolidity/analysis/ControlFlowAnalyzer.cpp

```diff
@@ -33,131 +33,122 @@ bool ControlFlowAnalyzer::visit(FunctionDefinition const& _function)
 	{
-		auto node = nodesToTraverse.top();
+		auto currentNode = nodesToTraverse.top();
 		nodesToTraverse.pop();
 
-		auto& unassignedAtNode = unassigned[node];
-
-		if (node->block.returnStatement != nullptr)
-			if (node->block.returnStatement->expression())
-				unassignedAtNode.clear();
-		if (!unassignedAtNode.empty())
+		auto& nodeInfo = nodeInfos[currentNode];
+		for (auto const& variableOccurrence: currentNode->variableOccurrences)
 		{
-			// kill all return values to which a value is assigned
-			for (auto const* variableDeclaration: variablesAssignedInNode(node))
-				unassignedAtNode.erase(variableDeclaration);
-
-			// kill all return values referenced in inline assembly
-			// a reference is enough, checking whether there actually was an assignment might be overkill
-			for (auto assembly: node->block.inlineAssemblyStatements)
-				for (auto const& ref: assembly->annotation().externalReferences)
-					if (auto variableDeclaration = dynamic_cast<VariableDeclaration const*>(ref.second.declaration))
-						unassignedAtNode.erase(variableDeclaration);
+			switch (variableOccurrence.kind())
+			{
+				case VariableOccurrence::Kind::Assignment:
+					nodeInfo.unassignedVariables.erase(&variableOccurrence.declaration());
+					break;
+				case VariableOccurrence::Kind::InlineAssembly:
+					// We consider all variable referenced in inline assembly as assignments.
+					// So far any reference is enough, but we might want to check whether
+					// there actually was an assignment in the future.
+					nodeInfo.unassignedVariables.erase(&variableOccurrence.declaration());
+					break;
+				case VariableOccurrence::Kind::Access:
+					if (nodeInfo.unassignedVariables.count(&variableOccurrence.declaration()))
+					{
+						if (variableOccurrence.declaration().type()->dataStoredIn(DataLocation::Storage))
+							// Merely store the unassigned access. We do not generate an error right away, since this
+							// path might still always revert. It is only an error if this is propagated to the exit
+							// node of the function (i.e. there is a path with an uninitialized access).
+							nodeInfo.uninitializedVariableAccesses.insert(&variableOccurrence);
+					}
+					break;
+				case VariableOccurrence::Kind::Declaration:
+					nodeInfo.unassignedVariables.insert(&variableOccurrence.declaration());
+					break;
+			}
 		}
 
-		for (auto const& exit: node->exits)
-		{
-			auto& unassignedAtExit = unassigned[exit];
-			auto oldSize = unassignedAtExit.size();
-			unassignedAtExit.insert(unassignedAtNode.begin(), unassignedAtNode.end());
-			// (re)traverse an exit, if we are on a path with new unassigned return values to consider
-			// this will terminate, since there is only a finite number of unassigned return values
-			if (unassignedAtExit.size() > oldSize)
+		// Propagate changes to all exits and queue them for traversal, if needed.
+		for (auto const& exit: currentNode->exits)
+			if (nodeInfos[exit].propagateFrom(nodeInfo))
```

Comment:
It looks to me as if (probably depending on the control flow graph) this might not terminate in some situations: In this function call, you add new variables to `unassignedVariables` and when you process the node, you remove stuff again - this might oszillate.
What about having pairs of sets: unassignedVariables that come in from other nodes and then the set of unassigned variables at the end of this node. You only re-process if "unassigned variables at entry" grows and of course you propagate from the "unassigned variables at exit".

Reply:
Let me double check, but you may be right and we may need to split the set into two.

Reply:
Yes, I'll change it for clarity. It should be apparent that the split doesn't actually change anything, but it'll be easier to see that it terminates.